In [1]:
# Import the dependencies
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

In [2]:
# Read the CSV into a DataFrame
df = pd.read_csv("/content/sales.csv")
df.tail(50)

PageVisitsCat1  PageVisitDurationCat1  PageVisitsCat2  \
12233             1.0              73.750000             0.0   
12234             1.0              88.000000             3.0   
12235             0.0               0.000000             0.0   
12236             0.0               0.000000             0.0   
12237             0.0               0.000000             0.0   
12238             6.0             369.333333             2.0   
12239             0.0               0.000000             0.0   
12240             8.0             167.910714             6.0   
12241             3.0             100.000000             0.0   
12242             0.0               0.000000             0.0   
12243             0.0               0.000000             0.0   
12244             0.0               0.000000             0.0   
12245             0.0               0.000000             0.0   
12246             0.0               0.000000             2.0   
12247             0.0               0.000000             0.0   
12248             0.0               0.000000             3.0   
12249             0.0               0.000000             0.0   
12250             0.0               0.000000             0.0   
12251             0.0               0.000000             0.0   
12252             3.0               9.000000             0.0   
12253             0.0               0.000000             0.0   
12254             0.0               0.000000             0.0   
12255             0.0               0.000000             0.0   
12256             0.0               0.000000             0.0   
12257             0.0               0.000000             0.0   
12258             0.0               0.000000             0.0   
12259             0.0               0.000000             0.0   
12260             2.0             305.125000             3.0   
12261             1.0              19.000000             0.0   
12262             3.0              54.000000             0.0   
12263             1.0               5.000000             0.0   
12264             1.0               0.000000             2.0   
12265             7.0             150.357143             1.0   
12266             3.0              16.000000             3.0   
12267             0.0               0.000000             0.0   
12268             0.0               0.000000             0.0   
12269             0.0               0.000000             0.0   
12270             3.0             100.500000             0.0   
12271             0.0               0.000000             0.0   
12272             0.0               0.000000             0.0   
12273             0.0               0.000000             0.0   
12274             0.0               0.000000             0.0   
12275             6.0              76.250000             0.0   
12276             2.0              64.750000             0.0   
12277             0.0               0.000000             1.0   
12278             3.0             145.000000             0.0   
12279             0.0               0.000000             0.0   
12280             0.0               0.000000             0.0   
12281             4.0              75.000000             0.0   
12282             0.0               0.000000             0.0   

       PageVisitDurationCat2  PageVisitsCat3  PageVisitDurationCat3  \
12233               0.000000            45.0            2674.339216   
12234             540.333333            57.0            1656.410714   
12235               0.000000           125.0            7453.766667   
12236               0.000000            37.0             854.550000   
12237               0.000000            45.0            2650.416667   
12238             225.500000           133.0            3918.363736   
12239               0.000000             4.0             197.416667   
12240             547.750000           111.0            6340.152381   
12241               0.000000            27.0             730.791667   
12242               0.000000    

In [3]:
# List the columns
for col in df.columns:
    print(col)

PageVisitsCat1
PageVisitDurationCat1
PageVisitsCat2
PageVisitDurationCat2
PageVisitsCat3
PageVisitDurationCat3
BounceRates
ExitRates
PageValues
HowCloseToSpecialDay
Month
VisitorType
Weekend
Revenue


### Description of columns:

**PageVisitsCatX:** Number of pages visited by a visitor. Cat1 is administrative, Cat2 is informational, and Cat3 is product related.

**PageVisitDurationCatX:** How long a visitor stayed on the category of pages.

**BounceRates:** Percentage of visitors who landed and exited a page.

**ExitRates:** Percentage of visitors who left the site from that page.

**PageValues:** A measurement of a page's contribution to a sale.

**HowCloseToSpecialDay:** How close is browsing date to a special day or a holiday, such as Valentine's day? Higher numbers are closer.

**Month:** Month of visit.

**VisitorType:** Is the visitor a new or returning visitor?

**Weekend:** Did the visit occur on a weekend day?

**Revenue:** Did the visit conclude in a sale?

## Perform Data Analysis

In [4]:
# Retrieve the number visits for each month.
Number_visit = df.groupby(['PageVisitsCat1', 'PageVisitsCat2', 'PageVisitsCat3']).Month.count().reset_index()
Number_visit

PageVisitsCat1  PageVisitsCat2  PageVisitsCat3  Month
0                0.0             0.0             0.0      5
1                0.0             0.0             1.0    547
2                0.0             0.0             2.0    370
3                0.0             0.0             3.0    331
4                0.0             0.0             4.0    281
...              ...             ...             ...    ...
3126            24.0             5.0            34.0      1
3127            24.0             6.0           178.0      1
3128            24.0             6.0           189.0      1
3129            26.0             9.0           183.0      1
3130            27.0             2.0           584.0      1

[3131 rows x 4 columns]

In [5]:
# Retrieve the number of visits that resulted in a purchase or not.
purchase_not = df.groupby(['PageVisitsCat3']).Revenue.sum().reset_index()
purchase_not

PageVisitsCat3  Revenue
0               0.0        6
1               1.0       13
2               2.0       20
3               3.0       25
4               4.0       18
..              ...      ...
306           518.0        0
307           534.0        1
308           584.0        0
309           686.0        0
310           705.0        0

[311 rows x 2 columns]

In [6]:
# What percentage of visits resulted in a purchase?
df['purchase'] = df.Revenue.notnull()
per_purchase = df.groupby(['PageVisitsCat3']).purchase.apply(lambda x:100 * x / float(x.sum()))
per_purchase

0        0.168919
1        0.217391
2        0.217391
3        0.303030
4        0.458716
           ...   
12278    1.666667
12279    0.263158
12280    0.253165
12281    0.370370
12282    0.219298
Name: purchase, Length: 12283, dtype: float64

In [7]:
# Filter the DataFrame for all purchases.
df.filter(["PageVisitsCat3","Month", "purchase"]).head()

PageVisitsCat3 Month  purchase
0             1.0   Feb      True
1             2.0   Feb      True
2             2.0   Feb      True
3            10.0   Feb      True
4            19.0   Feb      True

In [8]:
# What is the number of visits for each month that resulted in a purchase?
Num_vis_purchase = df.groupby(['PageVisitsCat3', 'purchase']).Month.count().reset_index()
Num_vis_purchase

PageVisitsCat3  purchase  Month
0               0.0      True     35
1               1.0      True    592
2               2.0      True    460
3               3.0      True    456
4               4.0      True    400
..              ...       ...    ...
306           518.0      True      1
307           534.0      True      1
308           584.0      True      1
309           686.0      True      1
310           705.0      True      1

[311 rows x 3 columns]

In [9]:
# Which month had the greatest number of visits where a purchase was made?
Max_vst = df.groupby(['PageVisitsCat3', 'purchase']).Month.max().reset_index()
Max_vst

PageVisitsCat3  purchase Month
0               0.0      True   Sep
1               1.0      True   Sep
2               2.0      True   Sep
3               3.0      True   Sep
4               4.0      True   Sep
..              ...       ...   ...
306           518.0      True   Oct
307           534.0      True   Nov
308           584.0      True   Nov
309           686.0      True   Aug
310           705.0      True   May

[311 rows x 3 columns]

In [10]:
# How many visits were from returning visitors?

df_ret = df.groupby(['PageVisitsCat1', 'PageVisitsCat2', 'PageVisitsCat3'])['VisitorType'].apply(lambda x: (x=='Returning_Visitor').count()).reset_index(name='Returning_Visitor')
df_ret

PageVisitsCat1  PageVisitsCat2  PageVisitsCat3  Returning_Visitor
0                0.0             0.0             0.0                  5
1                0.0             0.0             1.0                547
2                0.0             0.0             2.0                370
3                0.0             0.0             3.0                331
4                0.0             0.0             4.0                281
...              ...             ...             ...                ...
3126            24.0             5.0            34.0                  1
3127            24.0             6.0           178.0                  1
3128            24.0             6.0           189.0                  1
3129            26.0             9.0           183.0                  1
3130            27.0             2.0           584.0                  1

[3131 rows x 4 columns]

In [11]:
# How many visits were from new visitors?
new_vis = df.groupby(['PageVisitsCat1', 'PageVisitsCat2', 'PageVisitsCat3'])['VisitorType'].apply(lambda x: (x=='New_Visitor').count()).reset_index(name='New_Visitor')
new_vis

PageVisitsCat1  PageVisitsCat2  PageVisitsCat3  New_Visitor
0                0.0             0.0             0.0            5
1                0.0             0.0             1.0          547
2                0.0             0.0             2.0          370
3                0.0             0.0             3.0          331
4                0.0             0.0             4.0          281
...              ...             ...             ...          ...
3126            24.0             5.0            34.0            1
3127            24.0             6.0           178.0            1
3128            24.0             6.0           189.0            1
3129            26.0             9.0           183.0            1
3130            27.0             2.0           584.0            1

[3131 rows x 4 columns]

In [12]:
# How many visits took place on a weekday? And, on the weekend?
new_vis = df.groupby(['PageVisitsCat1', 'PageVisitsCat2', 'PageVisitsCat3'])['Weekend'].apply(lambda x: ((x=='True') & (x=='False')).count()).reset_index(name='Week_vst')
new_vis

PageVisitsCat1  PageVisitsCat2  PageVisitsCat3  Week_vst
0                0.0             0.0             0.0         5
1                0.0             0.0             1.0       547
2                0.0             0.0             2.0       370
3                0.0             0.0             3.0       331
4                0.0             0.0             4.0       281
...              ...             ...             ...       ...
3126            24.0             5.0            34.0         1
3127            24.0             6.0           178.0         1
3128            24.0             6.0           189.0         1
3129            26.0             9.0           183.0         1
3130            27.0             2.0           584.0         1

[3131 rows x 4 columns]

## Perform Logistic Regression 

In [24]:
# Convert categorical variables to binary variables 
df2 = pd.get_dummies(df)
df2

PageVisitsCat1  PageVisitDurationCat1  PageVisitsCat2  \
0                 0.0                    0.0             0.0   
1                 0.0                    0.0             0.0   
2                 0.0                    0.0             0.0   
3                 0.0                    0.0             0.0   
4                 0.0                    0.0             0.0   
...               ...                    ...             ...   
12278             3.0                  145.0             0.0   
12279             0.0                    0.0             0.0   
12280             0.0                    0.0             0.0   
12281             4.0                   75.0             0.0   
12282             0.0                    0.0             0.0   

       PageVisitDurationCat2  PageVisitsCat3  PageVisitDurationCat3  \
0                        0.0             1.0               0.000000   
1                        0.0             2.0              64.000000   
2                        0.0             2.0               2.666667   
3                        0.0            10.0             627.500000   
4                        0.0            19.0             154.216667   
...                      ...             ...                    ...   
12278                    0.0            53.0            1783.791667   
12279                    0.0             5.0             465.750000   
12280                    0.0             6.0             184.250000   
12281                    0.0            15.0             346.000000   
12282                    0.0             3.0              21.250000   

       BounceRates  ExitRates  PageValues  HowCloseToSpecialDay  ...  \
0         0.200000   0.200000    0.000000                   0.0  ...   
1         0.000000   0.100000    0.000000                   0.0  ...   
2         0.050000   0.140000    0.000000                   0.0  ...   
3         0.020000   0.050000    0.000000                   0.0  ...   
4         0.015789   0.024561    0.000000                   0.0  ...   
...            ...        ...         ...                   ...  ...   
12278     0.007143   0.029031   12.241717                   0.0  ...   
12279     0.000000   0.021333    0.000000                   0.0  ...   
12280     0.083333   0.086667    0.000000                   0.0  ...   
12281     0.000000   0.021053    0.000000                   0.0  ...   
12282     0.000000   0.066667    0.000000                   0.0  ...   

       Month_Jul  Month_June  Month_Mar  Month_May  Month_Nov  Month_Oct  \
0              0           0          0          0          0          0   
1              0           0          0          0          0          0   
2              0           0          0          0          0          0   
3              0           0          0          0          0          0   
4              0           0          0          0          0          0   
...          ...         ...        ...        ...        ...        ...   
12278          0           0          0          0          0          0   
12279          0           0          0          0          1          0   
12280          0           0          0          0          1          0   
12281          0           0          0          0          1          0   
12282          0           0          0          0          1          0   

       Month_Sep  VisitorType_New_Visitor  VisitorType_Other  \
0              0                        0                  0   
1              0                        0                  0   
2              0                        0                  0   
3              0                        0                  0   
4              0                        0                  0   
...          ...                      ...                ...   
12278          0                        0                  0   
12279          0                        0                  0   
12280          0               

In [25]:
# List the columns
for col in df2:
  print(col)

PageVisitsCat1
PageVisitDurationCat1
PageVisitsCat2
PageVisitDurationCat2
PageVisitsCat3
PageVisitDurationCat3
BounceRates
ExitRates
PageValues
HowCloseToSpecialDay
Weekend
Revenue
purchase
Month_Aug
Month_Dec
Month_Feb
Month_Jul
Month_June
Month_Mar
Month_May
Month_Nov
Month_Oct
Month_Sep
VisitorType_New_Visitor
VisitorType_Other
VisitorType_Returning_Visitor


In [26]:
# Separate the features and target variables.
y=df2["Revenue"]
X=df2.drop(columns="Revenue")

In [27]:
# Split the dataset into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test= train_test_split(X, 
                                                   y, 
                                                   random_state=1, 
                                                   stratify=y)

In [29]:
print(X_train.shape)
print(X_test.shape)
print(X.shape)

# Instantiate a logistic regression model
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(solver='lbfgs', random_state=1)
classifier

(9212, 25)
(3071, 25)
(12283, 25)


LogisticRegression(random_state=1)

In [30]:
# Train the model
classifier.fit(X_train, y_train)

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=1)

In [31]:
# Use the testing data to make predictions.
predictions = classifier.predict(X_test)

pd.DataFrame({"Prediction": predictions, "Actual": y_test}).reset_index(drop=True)

Prediction  Actual
0          False   False
1          False   False
2          False   False
3          False   False
4          False   False
...          ...     ...
3066       False   False
3067       False   False
3068       False    True
3069       False   False
3070       False   False

[3071 rows x 2 columns]

In [32]:
# Calculate the accuracy. 
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")


Training Data Score: 0.8797221016066001
Testing Data Score: 0.8870074894171279


In [33]:
# Print the confusion matrix.

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, predictions)


array([[2541,   53],
       [ 294,  183]])

In [34]:
# Print a classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

       False       0.90      0.98      0.94      2594
        True       0.78      0.38      0.51       477

    accuracy                           0.89      3071
   macro avg       0.84      0.68      0.72      3071
weighted avg       0.88      0.89      0.87      3071



## Overall Model Performance
----

- **Question:** Can the logistic regression model predict whether visiting a website will result in a purchase? 

- **Answer:**  Yes, because of the accuracy of the model, which is 89%.
